mm2src/rpc/rate_limiter.rs

```diff
@@ -0,0 +1,55 @@
+use crate::mm2::rpc::DispatcherError;
+use common::mm_ctx::from_ctx;
+use common::mm_ctx::MmArc;
+use common::mm_error::MmError;
+use derive_more::Display;
+use futures::lock::Mutex as AsyncMutex;
+use std::collections::HashMap;
+use std::net::{IpAddr, SocketAddr};
+use std::sync::Arc;
+
+pub const LIMIT_FAILED_REQUEST: usize = 10;
+pub type RateInfosRegistry = HashMap<IpAddr, usize>;
+
+#[derive(Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum RateLimitError {
+    #[display(fmt = "Rate Limit attempts left: {}", _0)]
+    NbAttemptsLeft(usize),
+}
+
+#[derive(Default)]
+pub struct RateLimitContext(AsyncMutex<RateInfosRegistry>);
+
+impl RateLimitContext {
+    pub fn from_ctx(ctx: &MmArc) -> Result<Arc<RateLimitContext>, String> {
+        Ok(try_s!(from_ctx(&ctx.rate_limit_ctx, move || {
+            Ok(RateLimitContext::default())
+        })))
+    }
+
+    pub async fn is_banned(&self, client_ip: IpAddr) -> bool {
+        let rate_infos = self.0.lock().await;
+        if let Some(limit) = rate_infos.get(&client_ip) {
+            return *limit == LIMIT_FAILED_REQUEST;
```

Comment:
If the limit somehow becomes larger than LIMIT_FAILED_REQUEST this will erroneously return false.

Reply:
Fixed but the case was not possible, each time before `process_rate_limit` is called `is_banned` is called before as a check.

```
let rate_limit_ctx = RateLimitContext::from_ctx(&ctx).unwrap();
    if rate_limit_ctx.is_banned(client.ip()).await {
        return ERR!("Your ip is banned.");
    }
    try_s!(auth(&req, &ctx, &client).await); ///< auth will process rate limit if required
```

Requests are not processed asynchronously. So the case that **limit** became > **LIMIT_FAILED_REQUEST** is just not possible.

Reply:
It might be a nitpick - it's not possible *now* of course and I understand this. However, we don't know who and how will change this code in the future turning `limit == LIMIT_FAILED_REQUEST` into a bug. It's better to be safe just in case.

Reply:
> It might be a nitpick - it's not possible _now_ of course and I understand this. However, we don't know who and how will change this code in the future turning `limit == LIMIT_FAILED_REQUEST` into a bug. It's better to be safe just in case.

Understand that, will apply >= everywhere then - thanks for the input.